### PR TITLE
Improve mypy typing for pyodide package

### PR DESCRIPTION
**Pull request by hoodmane**
The way we were handling the "routing" in `_core.py` (where in browser things are imported from `_pyodide_core` and otherwise things from `_pyodide._core_docs`) made mypy type a bunch of things as `Any`. This changes it to prevent mypy from understanding the imports from `_pyodide_core` so it will correctly type all of the methods with their types from `_pyodide._core_docs`.

This created a bunch of new type errors, so I also made various fixes to make things typecheck again.

### Checklists

- [ ] Add a [CHANGELOG](https://github.com/pyodide/pyodide/blob/main/docs/project/changelog.md) entry

### CODE REVIEW

File: src/py/_pyodide/_core_docs.py [hunk 1]
```diff
@@ -909,19 +954,70 @@ def create_proxy(
 # from python2js
 
 
+@overload
+def to_js(
+    obj: list[Any] | tuple[Any],
```
Comment (rth) [inline src/py/_pyodide/_core_docs.py:959]: It's unfortunate we have to repeat the rest of the annotations, just for this argument and output type, but I also don't see a way around it.

**Review by rth — APPROVED** [review]
Thanks LGTM.

At some point we need to see if there is a way to better format types in the documentation, as it's becoming not very readable
image

**Comment by hoodmane** [comment]
Thanks for the review @rth!

**Comment by hoodmane** [comment]
Agreed about the types in api docs. I'll try to look into improving it soon. For instance, it would be a huge improvement if instead of `collections.abc.Iterable` it could just say `Iterable`. Similarly, if we could render `Optional[T]` as something like `T?` and `Union[S,T]` as `S | T` this would be more readable. Also, if the types are really long, it would be good to start each argument on its own line.

**Comment by hoodmane** [comment]
It's also oddly inconsistent: See [throw](https://pyodide.org/en/latest/usage/api/python-api/ffi.html#pyodide.ffi.JsGenerator.throw) where the first argument uses `A | B` but the second argument uses `Union[A, B]`.

**Pull request merged**